src/composed/message/types.rs

```diff
@@ -433,34 +438,57 @@ impl Message {
                     SignatureType::Text
                 };
 
-                let signature_config = SignatureConfig::new_v4(
-                    Default::default(),
-                    typ,
-                    algorithm,
-                    hash_algorithm,
-                    hashed_subpackets,
-                    unhashed_subpackets,
-                );
-                (typ, signature_config.sign(key, key_pw, l.data())?)
+                let mut config = match sig_version {
+                    SignatureVersion::V4 => SignatureConfig::v4(typ, algorithm, hash_algorithm),
+                    SignatureVersion::V6 => {
+                        SignatureConfig::v6_with_rng(rng, typ, algorithm, hash_algorithm)?
+                    }
+                    _ => bail!("unsupported signature version {:?}", sig_version),
+                };
+                config.hashed_subpackets = hashed_subpackets;
+                config.unhashed_subpackets = unhashed_subpackets;
+
+                (typ, config.sign(key, key_pw, l.data())?)
             }
             _ => {
                 let typ = SignatureType::Binary;
-                let signature_config = SignatureConfig::new_v4(
-                    Default::default(),
-                    typ,
-                    algorithm,
-                    hash_algorithm,
-                    hashed_subpackets,
-                    unhashed_subpackets,
-                );
+
+                let mut config = match sig_version {
+                    SignatureVersion::V4 => SignatureConfig::v4(typ, algorithm, hash_algorithm),
+                    SignatureVersion::V6 => {
+                        SignatureConfig::v6_with_rng(rng, typ, algorithm, hash_algorithm)?
+                    }
+                    _ => bail!("unsupported signature version {:?}", sig_version),
+                };
+                config.hashed_subpackets = hashed_subpackets;
+                config.unhashed_subpackets = unhashed_subpackets;
 
                 let data = self.to_bytes()?;
-                let signature = signature_config.sign(key, key_pw, &data[..])?;
+                let signature = config.sign(key, key_pw, &data[..])?;
 
                 (typ, signature)
             }
         };
-        let ops = OnePassSignature::from_details(typ, hash_algorithm, algorithm, key_id);
+
+        let ops = match sig_version {
+            SignatureVersion::V4 => {
+                OnePassSignature::from_details_v3(typ, hash_algorithm, algorithm, key_id)
+            }
```

Comment:
should we check for the matching config version here too?

Reply:
We create the `SignatureConfig` in the block above this one. The versions for both packets are chosen based on the key version. I don't see how the `SignatureConfig` and `Ops` versions could fall out of sync, here.

But maybe I'm misreading what you are suggesting?

Fwiw: v6 keys can only make v6 `Signature`/`Ops` packets (and v4 keys must always make v4 `Signature` and v3 `Ops` packets). So picking the versions is nicely straightforward.

Reply:
After staring at this code for a while, I think it's clearer to `match` directly on the signer key version, when picking signature versions. See a45b2dff72b2e984268bcfd8f3195bcb392b53f2